go/vt/vttablet/tabletmanager/vreplication/vcopier.go

```diff
@@ -0,0 +1,293 @@
+	defer vc.vr.dbClient.Rollback()
+
+	log.Infof("Copying table %s, lastpk: %v", tableName, copyState[tableName])
+
+	plan, err := buildReplicatorPlan(vc.vr.source.Filter, vc.vr.tableKeys, nil)
+	if err != nil {
+		return err
+	}
+
+	initialPlan, ok := plan.TargetTables[tableName]
+	if !ok {
+		return fmt.Errorf("plan not found for table: %s, curret plans are: %#v", tableName, plan.TargetTables)
+	}
+
+	vsClient, err := tabletconn.GetDialer()(vc.vr.sourceTablet, grpcclient.FailFast(false))
+	if err != nil {
+		return fmt.Errorf("error dialing tablet: %v", err)
+	}
+	defer vsClient.Close(ctx)
+
+	ctx, cancel := context.WithTimeout(ctx, copyTimeout)
+	defer cancel()
+
+	target := &querypb.Target{
+		Keyspace:   vc.vr.sourceTablet.Keyspace,
+		Shard:      vc.vr.sourceTablet.Shard,
+		TabletType: vc.vr.sourceTablet.Type,
+	}
+
+	var lastpkpb *querypb.QueryResult
+	if lastpkqr := copyState[tableName]; lastpkqr != nil {
+		lastpkpb = sqltypes.ResultToProto3(lastpkqr)
+	}
+
+	var pkfields []*querypb.Field
+	var updateCopyState *sqlparser.ParsedQuery
+	err = vsClient.VStreamRows(ctx, target, initialPlan.SendRule.Filter, lastpkpb, func(rows *binlogdatapb.VStreamRowsResponse) error {
+		select {
+		case <-ctx.Done():
+			return io.EOF
+		default:
+		}
+		if vc.tablePlan == nil {
+			if len(rows.Fields) == 0 {
+				return fmt.Errorf("expecting field event first, got: %v", rows)
+			}
+			if err := vc.fastForward(ctx, copyState, rows.Gtid); err != nil {
+				return err
+			}
+			fieldEvent := &binlogdatapb.FieldEvent{
+				TableName: initialPlan.SendRule.Match,
+				Fields:    rows.Fields,
+			}
+			vc.tablePlan, err = plan.buildExecutionPlan(fieldEvent)
+			if err != nil {
+				return err
+			}
+			pkfields = rows.Pkfields
+			buf := sqlparser.NewTrackedBuffer(nil)
+			buf.Myprintf("update _vt.copy_state set lastpk=%a where vrepl_id=%s and table_name=%s", ":lastpk", strconv.Itoa(int(vc.vr.id)), encodeString(tableName))
+			updateCopyState = buf.ParsedQuery()
+		}
+		if len(rows.Rows) == 0 {
+			return nil
+		}
+		query, err := vc.tablePlan.generateBulkInsert(rows)
+		if err != nil {
+			return err
+		}
+		var buf bytes.Buffer
+		err = proto.CompactText(&buf, &querypb.QueryResult{
+			Fields: pkfields,
+			Rows:   []*querypb.Row{rows.Lastpk},
+		})
+		if err != nil {
+			return err
+		}
+		bv := map[string]*querypb.BindVariable{
+			"lastpk": {
+				Type:  sqltypes.VarBinary,
+				Value: buf.Bytes(),
+			},
+		}
+		updateState, err := updateCopyState.GenerateQuery(bv, nil)
+		if err != nil {
+			return err
+		}
+		if err := vc.vr.dbClient.Begin(); err != nil {
+			return err
+		}
+		if _, err := vc.vr.dbClient.ExecuteFetch(query, 0); err != nil {
+			return err
+		}
+		if _, err := vc.vr.dbClient.ExecuteFetch(updateState, 0); err != nil {
+			return err
+		}
+		if err := vc.vr.dbClient.Commit(); err != nil {
+			return err
+		}
+		return nil
+	})
+	// If there was a timeout, return without an error.
+	select {
+	case <-ctx.Done():
+		return nil
```

Comment:
Should this explicitly verify that ctx.Err() is DeadlineExceeded before suppressing the error?